File: O(n!)/GerarPermutacoes.java
import java.util.Arrays;

public class GerarPermutacoes {
    // Função auxiliar para trocar elementos
    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    public static void gerarPermutacoes(int[] arr, int inicio, int fim) {
        // Se inicio for igual a fim, imprimimos a permutação atual
        if (inicio == fim) {
            System.out.println(Arrays.toString(arr));
        } else {
            for (int i = inicio; i <= fim; i++) {
                // Troca
                swap(arr, inicio, i);
                
                // Recursão
                gerarPermutacoes(arr, inicio + 1, fim);
                
                // Backtrack (Desfaz a troca)
                swap(arr, inicio, i);
            }
        }
    }

    public static void main(String[] args) {
        int[] dados = {1, 2, 3};
        System.out.println("Gerando permutações:");
        gerarPermutacoes(dados, 0, dados.length - 1);
    }
}

File: O(n!)/gerarPermutacoes.c
#include <stdio.h>

// Função para trocar os valores de dois ponteiros
void trocar(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

void gerarPermutacoes(int *arr, int inicio, int fim) {
    int i;
    // Se chegamos ao fim, imprime a permutação
    if (inicio == fim) {
        for (i = 0; i <= fim; i++) {
            printf("%d ", arr[i]);
        }
        printf("\n");
    } else {
        for (i = inicio; i <= fim; i++) {
            // Troca o elemento atual pelo elemento do índice i
            trocar((arr + inicio), (arr + i));
            
            // Chama recursivamente
            gerarPermutacoes(arr, inicio + 1, fim);
            
            // Backtrack: desfaz a troca
            trocar((arr + inicio), (arr + i));
        }
    }
}

int main() {
    int dados[] = {1, 2, 3};
    int n = sizeof(dados) / sizeof(dados[0]);
    
    printf("Gerando permutacoes:\n");
    gerarPermutacoes(dados, 0, n - 1);
    
    return 0;
}